utils/yoke/src/cart.rs

```diff
@@ -0,0 +1,117 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Yoke, Yokeable};
+use std::rc::Rc;
+use std::sync::Arc;
+
+/// A [`Cart`] is a type that is acceptable as backing storage in a [`Yoke`].
+///
+/// The essential invariant that must be maintained by implementors is that `Self::Inner` references
+/// obtained from this type via `Self::get_inner()` must continue to be valid for the
+/// duration of this type provided that this type is never accessed via `&mut` references.
+///
+/// For example, `Vec<u8>` is a valid [`Cart`], however `Vec<RefCell<u8>>` is not, because
+/// in the latter type it is possible to use interior mutation to change the data.
+///
+/// In general, this means "no interior mutability", though interior mutability can be fine
+/// if the interior mutability is to something that does not affect references. For example,
+/// `Rc<[u8]>` does have interior mutability in the refcount, but you cannot get references
+/// to it, so it's fine. On the other hand, `Weak<Box<[u8]>>` cannot be a valid [`Cart`] because
+/// it is possible for the backing buffer to be cleaned up without warning.
+///
+/// Common [`Cart`] types to use with [`Yoke`] are ones wrapping `[u8]` or `str`
+/// (`Box<[u8]>`, `Rc<[u8]>`, etc) since those are typical inputs to zero-copy
+/// deserialization and parsing.
+///
+/// Typically the [`Cart`] trait will be implemented by smart pointers whilst [`Cartable`] is used
+/// as a helper to talk about types that have the same property provided they are not moved, sucn that
+/// [`Cart`] can be implemented on wrappers that can maintain this property even if they are moved.
+pub unsafe trait Cart {
+    type Inner: ?Sized;
+    /// Get the inner data
+    fn get_inner(&self) -> &Self::Inner;
+}
+
+/// [`Cartable`] is a helper trait for implementng [`Cart`]. It has a similar invariant:
+/// all references obtained from this type must continue to be valid for the lifetime
+/// of this type provided that this type is never moved or accessed via an `&mut` reference.
+///
+/// Essentially, this means "no interior mutability", however interior mutability which cannot be
+/// seen from the outside is fine.
```

Comment:
I think it is important to clarify what "seen" means here, since it is the essential difference between Rc and Weak described above.